### ISSUE TITLE
Unexpected behaviour of ulong -> int cast

### ISSUE THREAD

**Issue by Zastai** [dotnet/runtime#46529]
### Casts from `ulong` to `int` changed behaviour in .NET Core 3.0

We are updating an existing codebase to target multiple frameworks (`net472` plus the active LTS version of .NET).
We ran into odd behaviour with code that interpreted bytes as a big-endian two's complement integer.
It turns out conversions from `ulong` to `int` behave differently depending on whether it's one big expression, or uses an intermediate variable.

```cs
var bytes = new byte[] { 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xfe };

var direct = (int) ((((ulong)(bytes[0])) << 56) |
                    (((ulong)(bytes[1])) << 48) |
                    (((ulong)(bytes[2])) << 40) |
                    (((ulong)(bytes[3])) << 32) |
                    (((ulong)(bytes[4])) << 24) |
                    (((ulong)(bytes[5])) << 16) |
                    (((ulong)(bytes[6])) << 8) |
                    (((ulong)(bytes[7])) << 0));
var intermediate = (((ulong)(bytes[0])) << 56) |
                   (((ulong)(bytes[1])) << 48) |
                   (((ulong)(bytes[2])) << 40) |
                   (((ulong)(bytes[3])) << 32) |
                   (((ulong)(bytes[4])) << 24) |
                   (((ulong)(bytes[5])) << 16) |
                   (((ulong)(bytes[6])) << 8) |
                   (((ulong)(bytes[7])) << 0);
var cast = (int) intermediate;

Console.WriteLine($"Bytes: {BitConverter.ToString(bytes)}");
Console.WriteLine($"Direct Conversion: {direct}");
Console.WriteLine($"Conversion Using Intermediate: {intermediate} -> {cast}");
```

(I'm sure this code could be written much better - but it's the real-world code that showed the issue for us.)
The expected result is -2, regardless of which approach is taken.

### Configuration

After detecting this when targeting `netcoreapp3.1`, I set up the above example as a console app, with

```xml
    <TargetFrameworks>net40;net472;netcoreapp2.1;netcoreapp2.2;netcoreapp3.0;netcoreapp3.1;net5.0</TargetFrameworks>
    <CheckEolTargetFramework>false</CheckEolTargetFramework>
```

This shows:

| Framework | Direct Result | Indirect Result |
|---|---|---|
| net40 | -2 | -2 |
| net472 | -2 | -2 |
| netcoreapp2.1 | -2 | -2 |
| netcoreapp2.2 | -2 | -2 |
| netcoreapp3.0 | -1 | -2 |
| netcoreapp3.1 | -1 | -2 |
| net5.0 | -1 | -2 |

Debug/Release configuration makes no difference.
`intermediate` always contains 18446744073709551614.

### Regression?

The behaviour in .NET Framework and .NET Core 2.x seems sensible.
The new behaviour is decidedly odd and unexpected, so I would class this as a regression.

### Other information

While there is a "workaround" of sorts (using a variable to store the computed ulong, which is also useful for debugging), it's a very subtle issue, and there is no diagnostic or runtime error, making places where such a workaround should be applied non-trivial to find.

**Comment by huoyaoyuan** [comment]
Explored with [sharplab](https://sharplab.io/#v2:EYLgxg9gTgpgtADwGwBYA0AXEBDAzgWwB8ABAJgEYBYAKGIGYACMhgYQYG8aHunGBLAHYYGAET6wwGABTAAnhhgBtALoM5C3AEouPTtR4GGAN2xQGAE3ExJDALwMpgjJodS3AVwA2EAQHNNMvIwuIoADMqaLgA8UQwArEguhDqGqWluUl4+/oEaiuQR0bEoABxJKWmV3BlZfgHqwYqkhQwxDCih5fpVPTXedbmNdC1tdKRdPb0e/TkNISgjsaQoE5OVfdn1QSFxiwzkiQzJ3WupGwNzikh7ZUcVpzzns9uKAOx7nZoA3Pc8xK8WKySH4nBgAXwq9AYTgYAEkBOYYAooPg+JZsApBio1NttKC9JUTGYnDAUTB0Qo7K5MjMtnlwpFWrEEqsHo9pptBiECoy2qVWWzqbVnnlmrzip07qC2U86UM9mMBTKORcXgtxQxlkqHrKuYpdhqDtrTrrLtcNbdjoL2TTOZd3hrOiDCaYGGA8MJ7I4hC4SWSKTBnWl/m6PUHuBDQZDGMQUAwALJSPEGAlpIk4jRUgQwADuGaUqnYDFCCAAZqW0MWyxWq+XKyW67Waw3m9X62WYOCQTR05YJJ6HE4XBlbar6XsWVLrTbhXLuXt+VPp0LaXqxUV2pKrcvTS9hhrFUvp7u8uqN1qj9aT40DRujZfBdeQuaN5bfjqVSLGg6N58Qem/Xwck+AxTsvU/OcwgnQ5tyvCC9R5DdF1gx94MudcmU3Y01ifRR9w3Q8UOVUcv3mPYLyIj8SMg29MPvSiTTQl4X0wt9pSo2c9R/TCnR7V13VwAdvWcaEhFJICA27WhyAATikAASAAiAAhbYQA4ZS+AwFgfCMUlkQAOgAFQgABlDAoEEXwuU0MFFO+GhiFkhTFLEftWF00lcD4Hx1PYPtrAwOyHOkuSlJ0gQ9KgbyfAYABVby/DhMT/RAhQ/MA4DQLBBg4AAPg4ASgvs8NwRoMEgA).

When I changing the input values from `bytes`, I can't predict it's behavior. This looks like a JIT bug that only reproduces for such complex expression.

**Comment by EgorBo** [comment]
It does look like a bug (reproduces on `[netcoreapp3.0 .. net6]`), a minimal repro:
```csharp
using System;

public unsafe class Test
{
    public static void Main()
    {
        byte p = 0xFF;
        Foo(&p);
        Console.ReadKey();
    }

    static void Foo(byte* bytes)
    {
        ulong intermediate = ((ulong)bytes[0]) << 56;
        Console.WriteLine((int)(((ulong)bytes[0]) << 56));
        Console.WriteLine((int)intermediate);
    }
}
```
Output:
```
-16777216  (0 is expected)
0
```
Codegen for `Foo`:
```asm
G_M27899_IG01:
       push     rsi
       sub      rsp, 32
G_M27899_IG02:
       movzx    rcx, byte  ptr [rcx]
       mov      rsi, rcx
       shl      rsi, 56
       shl      ecx, 56 ;;; rcx, 56 ?
       call     System.Console:WriteLine(int)
       mov      ecx, esi
G_M27899_IG03:
       add      rsp, 32
       pop      rsi
       jmp      System.Console:WriteLine(int)
```
/cc @dotnet/jit-contrib

**Comment by EgorBo** [comment]
Potential fix (works for the repros): https://github.com/EgorBo/runtime-1/commit/7a8d73898de08384294a75be18e1e68230a10998

**Comment by AndyAyersMS** [comment]
cc @dotnet/jit-contrib 

This bit of code in morph has been problematic...

**Comment by Zastai** [comment]
So a hard-to-diagnose bug in code generation is being deferred for nearly a year, when there seems to be a fix available? Or am I misinterpreting the 6.0.0 milestone?

**Comment by huoyaoyuan** [comment]
Once it's mentioned, a fix would be very soon. I'd expect it to be fixed in 5.0.3.

**Comment by briansull** [comment]
I will take a look

**Comment by AndyAyersMS** [comment]
@Zastai we usually fix in master first, then work on porting the fix back to previous releases.

**Comment by briansull** [comment]
@EgorBo You have a reasonable fox for this issue.
I will reassign this issue to you.

**Comment by briansull** [comment]
I will take this  back, as it requires a servicing fix

**Comment by JulieLeeMSFT** [comment]
@briansull reopening this for both 5.0 and 3.1 servicing. Please create new issues for servicing.

**Comment by light-traveller** [comment]
**NON-TECHNICAL**

*I've been using MS development products and platforms as a developer for almost two decades-unfortunately.*

It is way too surprising that such basic fundamental flaws (many I have seen in this and other repositories such as [string Equals StringComparison.OrdinalIgnoreCase bug (#45889)](https://github.com/dotnet/runtime/issues/45889)) do exist in the first place! Writing tests for cases like (https://github.com/dotnet/runtime/issues/45889) is not hard.

It seems that Microsoft relies more on the users in order to deliver a reliable product, rather on its internal resources. 
This issue, particularly in the recent years with the strategy of moving to open source world, is being far more visible.

**MS excels at releasing immature products and as is being seen, there is no intention to change this attitude/strategy.**

This is profoundly bad for such a big corporation. More than anyone can possibly think of.

**Comment by briansull** [comment]
Posted the fix for [release/3.1]
dotnet/coreclr#28146 [release/3.1] Backport the fix for dotnet/runtime issue 46529
